service/history/workflow/mutable_state_impl.go

```diff
@@ -6252,3 +6260,419 @@ func (ms *MutableStateImpl) RefreshExpirationTimeoutTask(ctx context.Context) er
+					return err
+				}
+			}
+		}
+	}
+
+	getSizeOfKey := func(key any) int {
+		val := reflect.ValueOf(key)
+		if val.Kind() == reflect.String {
+			return int(len(val.String()))
+		}
+		return int(reflect.TypeOf(key).Size())
+	}
+
+	for key, updated := range updatedSubStateMachine {
+		var existing V
+		if existing, ok := pendingInfos[key]; ok {
+			if CompareVersionedTransition(existing.GetLastUpdateVersionedTransition(), updated.GetLastUpdateVersionedTransition()) == 0 {
+				continue
+			}
+			ms.approximateSize -= existing.Size() + getSizeOfKey(key)
+		}
+		val := updated
+		if sanitizeFn != nil {
+			val = common.CloneProto(updated)
+			sanitizeFn(existing, val)
+		}
+		pendingInfos[key] = val
+		updateInfos[key] = val
+		ms.approximateSize += val.Size() + getSizeOfKey(key)
+		if postUpdateFn != nil {
+			postUpdateFn(val)
+		}
+	}
+	return nil
+}
+
+func (ms *MutableStateImpl) applyUpdatesToUpdateInfos(
+	updatedUpdateInfos map[string]*persistencespb.UpdateInfo,
+	isSnapshot bool,
+) {
+	if isSnapshot {
+		for updateID := range ms.executionInfo.UpdateInfos {
+			if _, ok := updatedUpdateInfos[updateID]; !ok {
+				ms.approximateSize -= ms.executionInfo.UpdateInfos[updateID].Size() + len(updateID)
+				delete(ms.executionInfo.UpdateInfos, updateID)
+			}
+		}
+	}
+	for updateID, ui := range updatedUpdateInfos {
+		if existing, ok := ms.executionInfo.UpdateInfos[updateID]; ok {
+			if CompareVersionedTransition(existing.GetLastUpdateVersionedTransition(), ui.GetLastUpdateVersionedTransition()) == 0 {
+				continue
+			}
+			ms.approximateSize -= existing.Size() + len(updateID)
+		} else {
+			ms.executionInfo.UpdateCount++
+		}
+		ms.executionInfo.UpdateInfos[updateID] = ui
+		ms.approximateSize += ui.Size() + len(updateID)
+	}
+}
+
+func (ms *MutableStateImpl) syncExecutionInfo(current *persistencespb.WorkflowExecutionInfo, incoming *persistencespb.WorkflowExecutionInfo) error {
+	doNotSyncFields := []interface{}{
+		&current.WorkflowTaskVersion,
+		&current.WorkflowTaskScheduledEventId,
+		&current.WorkflowTaskStartedEventId,
+		&current.WorkflowTaskRequestId,
+		&current.WorkflowTaskTimeout,
+		&current.WorkflowTaskAttempt,
+		&current.WorkflowTaskStartedTime,
+		&current.WorkflowTaskScheduledTime,
+		&current.WorkflowTaskOriginalScheduledTime,
+		&current.WorkflowTaskType,
+		&current.WorkflowTaskSuggestContinueAsNew,
+		&current.WorkflowTaskHistorySizeBytes,
+		&current.WorkflowTaskBuildId,
+		&current.WorkflowTaskBuildIdRedirectCounter,
+		&current.VersionHistories,
+		&current.ExecutionStats,
+		&current.LastFirstEventTxnId,
+		&current.ParentClock,
+		&current.CloseTransferTaskId,
+		&current.CloseVisibilityTaskId,
+		&current.CloseVisibilityTaskCompleted,
+		&current.WorkflowExecutionTimerTaskStatus,
+		&current.SubStateMachinesByType,
+		&current.StateMachineTimers,
+		&current.TaskGenerationShardClockTimestamp,
+		&current.UpdateInfos,
+	}
+	err := common.MergeProtoExcludingFields(current, incoming, doNotSyncFields...)
```

Comment:
nit: I am a little bit concerned with the reflection usage here TBH since this will be a quite hot code path that got invoked for every replication task. When you have time can you do a benchtest on this util function to see if it will become a bottleneck? Thanks.